src/sql/parts/notebook/notebookInput.ts

```diff
@@ -8,150 +8,189 @@
+export class NotebookEditorModel extends EditorModel {
 	private dirty: boolean;
 	private readonly _onDidChangeDirty: Emitter<void> = this._register(new Emitter<void>());
-	private _providerId: string;
-	private _standardKernels: IStandardKernelWithProvider[];
-	private _defaultKernel: azdata.nb.IKernelSpec;
 	constructor(public readonly notebookUri: URI,
-		private readonly handle: number,
-		private _isTrusted: boolean = false,
-		private saveHandler?: ModeViewSaveHandler,
-		provider?: string,
-		private _providers?: string[],
-		private _connectionProfileId?: string) {
-
+		private textEditorModel: TextFileEditorModel | UntitledEditorModel,
+		@INotebookService private notebookService: INotebookService
+	) {
 		super();
-		this.dirty = false;
-		this._providerId = provider;
-		this._standardKernels = [];
-	}
-
-	public get providerId(): string {
-		return this._providerId;
-	}
-
-	public set providerId(value: string) {
-		this._providerId = value;
-	}
-
-	public get providers(): string[] {
-		return this._providers;
-	}
-
-	public set providers(value: string[]) {
-		this._providers = value;
-	}
-
-	public get connectionProfileId(): string {
-		return this._connectionProfileId;
-	}
-
-	public get standardKernels(): IStandardKernelWithProvider[] {
-		return this._standardKernels;
-	}
-
-	public set standardKernels(value: IStandardKernelWithProvider[]) {
-		value.forEach(kernel => {
-			this._standardKernels.push({
-				connectionProviderIds: kernel.connectionProviderIds,
-				name: kernel.name,
-				notebookProvider: kernel.notebookProvider
-			});
-		});
-	}
-
-	public get defaultKernel(): azdata.nb.IKernelSpec {
-		return this._defaultKernel;
-	}
+		this._register(this.notebookService.onNotebookEditorAdd(notebook => {
+			if (notebook.id === this.notebookUri.toString()) {
+				// Hook to content change events
+				notebook.modelReady.then(() => {
+					this._register(notebook.model.contentChanged(e => this.updateModel()));
+					this._register(notebook.model.kernelChanged(e => this.updateModel()));
+				}, err => undefined);
+			}
+		}));
 
-	public set defaultKernel(kernel: azdata.nb.IKernelSpec) {
-		this._defaultKernel = kernel;
+		if (this.textEditorModel instanceof UntitledEditorModel) {
+			this._register(this.textEditorModel.onDidChangeDirty(e => this.setDirty(this.textEditorModel.isDirty())));
+		} else {
+			this._register(this.textEditorModel.onDidStateChange(e => this.setDirty(this.textEditorModel.isDirty())));
+		}
+		this.dirty = this.textEditorModel.isDirty();
 	}
 
-	get isTrusted(): boolean {
-		return this._isTrusted;
-	}
-
-	get onDidChangeDirty(): Event<void> {
-		return this._onDidChangeDirty.event;
+	public get contentString(): string {
+		let model = this.textEditorModel.textEditorModel;
+		return model.getValue();
 	}
 
 	get isDirty(): boolean {
-		return this.dirty;
+		return this.textEditorModel.isDirty();
 	}
 
 	public setDirty(dirty: boolean): void {
 		if (this.dirty === dirty) {
```

Comment:
Oh... isn't this where you meant to use the modelLoaded event? E.g. lock updates to dirty while loading model, then on modelLoaded you re-enable dirty updates? 
- Remember you're not actually firing the event, so in addition to registering for it we need to actually send it on load completing 😄

Reply:
Unresolving as I don't see this fixed yet in your current PR update...